core/blockchain_eip_3529_test.go

```diff
@@ -0,0 +1,169 @@
+}
+
+func TestSstoreGasPreEIP3529(t *testing.T) {
+	bytecode := []byte{
+		byte(vm.PUSH1), 0x3, // value
+		byte(vm.PUSH1), 0x3, // location
+		byte(vm.SSTORE), // Set slot[3] = 3
+	}
+	// 	// Expected gas is intrinsic +  2*pushGas + cold load (due to legacy tx) +  SstoreGas
+	expectedGasUsed := params.TxGas + 2*vm.GasFastestStep + params.ColdSloadCostEIP2929 + params.SstoreSetGasEIP2200
+	testGasUsage(t, preEIP3529Config(), ethash.NewFaker(), bytecode, nil, 60_000, expectedGasUsed)
+}
+
+func TestSelfDestructGasPostEIP3529(t *testing.T) {
+	bytecode := []byte{
+		byte(vm.PC),
+		byte(vm.SELFDESTRUCT),
+	}
+	// Expected gas is intrinsic +  pc + cold load (due to legacy tx) + SelfDestructGas
+	expectedGasUsed := params.TxGas + vm.GasQuickStep + params.ColdAccountAccessCostEIP2929 + params.SelfdestructGasEIP150
+	testGasUsage(t, bonehConfig(), ethash.NewFaker(), bytecode, nil, 60_000, expectedGasUsed)
+}
+
+func TestSstoreGasPostEIP3529(t *testing.T) {
+	bytecode := []byte{
+		byte(vm.PUSH1), 0x3, // value
+		byte(vm.PUSH1), 0x3, // location
+		byte(vm.SSTORE), // Set slot[3] = 3
+	}
+	// Expected gas is intrinsic +  2*pushGas + cold load (due to legacy tx) +  SstoreGas
+	expectedGasUsed := params.TxGas + 2*vm.GasFastestStep + params.ColdSloadCostEIP2929 + params.SstoreSetGasEIP2200
+	testGasUsage(t, bonehConfig(), ethash.NewFaker(), bytecode, nil, 60_000, expectedGasUsed)
+}
+
+func TestSstoreModifyGasPostEIP3529(t *testing.T) {
+	bytecode := []byte{
+		byte(vm.PUSH1), 0x3, // value
+		byte(vm.PUSH1), 0x1, // location
+		byte(vm.SSTORE), // Set slot[1] = 3
+	}
+	// initialize contract storage
+	initialStorage := make(map[common.Hash]common.Hash)
+	// Populate two slots
+	initialStorage[common.HexToHash("01")] = common.HexToHash("01")
+	initialStorage[common.HexToHash("02")] = common.HexToHash("02")
+	// Expected gas is intrinsic +  2*pushGas + cold load (due to legacy tx) + SstoreReset (a->b such that a,b!=0)
+	expectedGasUsed := params.TxGas + 2*vm.GasFastestStep + params.ColdSloadCostEIP2929 + (params.SstoreResetGasEIP2200 - params.ColdSloadCostEIP2929)
+	testGasUsage(t, bonehConfig(), ethash.NewFaker(), bytecode, initialStorage, 60_000, expectedGasUsed)
+}
+
+func TestSstoreClearGasPostEIP3529(t *testing.T) {
+	bytecode := []byte{
+		byte(vm.PUSH1), 0x0, // value
+		byte(vm.PUSH1), 0x1, // location
+		byte(vm.SSTORE), // Set slot[1] = 0
+	}
+	// initialize contract storage
+	initialStorage := make(map[common.Hash]common.Hash)
+	// Populate two slots
+	initialStorage[common.HexToHash("01")] = common.HexToHash("01")
+	initialStorage[common.HexToHash("02")] = common.HexToHash("02")
+
+	// Expected gas is intrinsic +  2*pushGas + cold load (due to legacy tx) + SstoreReset (a->b such that a,b!=0) - sstoreClearGasRefund
+	expectedGasUsage := params.TxGas + 2*vm.GasFastestStep + params.ColdSloadCostEIP2929 + (params.SstoreResetGasEIP2200 - params.ColdSloadCostEIP2929) - params.SstoreClearsScheduleRefundEIP3529
```

Comment:
Could you explain this a bit more?

Reply:
Ah, I notice I have a mistake in the comment. The modification is `a-> 0` where `a!=0`.

Essentially what happens is that running the bytecode you have gas costs for:
- intrinsic transaction gas
- the two `PUSH` operations
- Modifying a storage variable

and you have get a refund for:
- Setting a storage variable to 0 from some other value (the refund is `params.SstoreClearsScheduleRefundEIP3529`)